contrib/educational_decoder/zstd_decompress.c

```diff
@@ -0,0 +1,2168 @@
+    memcpy(streams->dst, literals, literals_len);
+    streams->dst += literals_len;
+    streams->dst_len -= literals_len;
+
+    total_output += literals_len;
+
+    ctx->current_total_output = total_output;
+}
+/******* END SEQUENCE EXECUTION ***********************************************/
+
+/******* OUTPUT SIZE COUNTING *************************************************/
+size_t traverse_frame(const frame_header_t *const header, const u8 *src,
+                      size_t src_len);
+
+/// Get the decompressed size of an input stream so memory can be allocated in
+/// advance.
+/// This is more complex than the implementation in the reference
+/// implementation, as this API allows for the decompression of multiple
+/// concatenated frames.
+size_t ZSTD_get_decompressed_size(const void *src, const size_t src_len) {
+  const u8 *ip = (const u8 *) src;
+  size_t ip_len = src_len;
+  size_t dst_size = 0;
+
+  // Each frame header only gives us the size of its frame, so iterate over all
+  // frames
+  while (ip_len > 0) {
+    if (ip_len < 4) {
+      INP_SIZE();
+    }
+
+    const u32 magic_number = read_bits_LE(ip, 32, 0);
+
+    ip += 4;
+    ip_len -= 4;
+    if (magic_number >= 0x184D2A50U && magic_number <= 0x184D2A5F) {
+        // skippable frame, this has no impact on output size
+        if (ip_len < 4) {
+            INP_SIZE();
+        }
+        const size_t frame_size = read_bits_LE(ip, 32, 32);
+
+        if (ip_len < 4 + frame_size) {
+            INP_SIZE();
+        }
+
+        // skip over frame
+        ip += 4 + frame_size;
+        ip_len -= 4 + frame_size;
+    } else if (magic_number == 0xFD2FB528U) {
+        // ZSTD frame
+        frame_header_t header;
+        parse_frame_header(&header, ip, ip_len);
+
+        if (header.frame_content_size == 0 && !header.single_segment_flag) {
+            // Content size not provided, we can't tell
+            return -1;
+        }
+
+        dst_size += header.frame_content_size;
+
+        // we need to traverse the frame to find when the next one starts
+        const size_t traversed = traverse_frame(&header, ip, ip_len);
+        ip += traversed;
+        ip_len -= traversed;
+    } else {
+        // not a real frame
+        ERROR("Invalid magic number");
+    }
+  }
+
+  return dst_size;
+}
+
+/// Iterate over each block in a frame to find the end of it, to get to the
+/// start of the next frame
+size_t traverse_frame(const frame_header_t *const header, const u8 *src,
+                      size_t src_len) {
+    const u8 *const src_beg = src;
+    const u8 *const src_end = src + src_len;
+    src += header->header_size;
+    src_len += header->header_size;
+
+    int last_block = 0;
+
+    do {
+        if (src + 3 > src_end) {
+            INP_SIZE();
+        }
+        // Parse the block header
+        last_block = src[0] & 1;
+        const int block_type = (src[0] >> 1) & 3;
+        const size_t block_len = read_bits_LE(src, 21, 3);
```

Comment:
Can you add a function for this, since this is copy/pasted from above?

Also, make an enum for `block_type`.